File: leet/rotate-image.ts
/**
 * https://leetcode.com/problems/rotate-image/description/
 */

export function rotate(matrix: number[][]): void {
  transpose(matrix);
  swap(matrix);
}

export function transpose(matrix: number[][]): void {
  const size = matrix.length;
  for (let i = 0; i < size; i++) {
    for (let j = i + 1; j < size; j++) {
      const temp = matrix[i][j];
      matrix[i][j] = matrix[j][i];
      matrix[j][i] = temp;
    }
  }
}

export function swap(matrix: number[][]): void {
  const size = matrix.length;
  const half = Math.floor(size / 2);
  for (let j = 0; j < half; j++) {
    for (let i = 0; i < size; i++) {
      const temp = matrix[i][j];
      matrix[i][j] = matrix[i][size - j - 1];
      matrix[i][size - j - 1] = temp;
    }
  }
}

File: returned/python/today3/rotate_image.py
# https://leetcode.com/problems/rotate-image/description/
from typing import List

class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        top = 0
        bottom = len(matrix) - 1
        left = 0
        right = bottom

        while left < right and top < bottom:
            i = left
            while i < right:
                tmp_right = matrix[top+i][right]
                matrix[top+i][right] = matrix[top][left+i]

                tmp_bottom = matrix[bottom][right-i]
                matrix[bottom][right-i] = tmp_right

                tmp_left = matrix[bottom-i][left]
                matrix[bottom-i][left] = tmp_bottom

                matrix[top][left+i] = tmp_left

                i += 1
            left += 1
            right -= 1
            top += 1
            bottom -= 1